sdk/tests/conformance2/textures/misc/tex-unpack-params-with-flip-y-and-premultiply-alpha.html

```diff
@@ -0,0 +1,511 @@
+  var padding = 0;
+  switch (alignment) {
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+      padding = (bytesPerPixel * width) % alignment;
+      if (padding > 0)
+        padding = alignment - padding;
+      return padding;
+    default:
+      testFailed("should not reach here");
+      return;
+  }
+}
+
+function computeImageSizes2D(width, height, testCase) {
+  // Assume RGBA8/UNSIGNED_BYTE
+  var bytesPerPixel = 4;
+  var actualWidth = testCase.rowLength == 0 ? width : testCase.rowLength;
+  var padding = calculatePaddingBytes(bytesPerPixel, testCase.alignment, actualWidth);
+  var bytesPerRow = actualWidth * bytesPerPixel + padding;
+  var bytesLastRow = bytesPerPixel * width;
+  var size = bytesPerRow * (height - 1) + bytesLastRow;
+  var skipSize = 0;
+  if (testCase.skipPixels > 0)
+    skipSize += bytesPerPixel * testCase.skipPixels;
+  if (testCase.skipRows > 0)
+    skipSize += bytesPerRow * testCase.skipRows;
+  return {size: size,
+          bytesPerRow: bytesPerRow,
+          bytesLastRow: bytesLastRow,
+          padding: padding,
+          skipSize: skipSize,
+          totalSize: size + skipSize};
+}
+
+function copyData(srcData, srcIndex, dstData, dstIndex, size, premultiply_alpha) {
+  for (var ii = 0; ii < size / 4; ++ii) {
+    var factor = 1.0;
+    if (premultiply_alpha)
+      factor = srcData[srcIndex + ii * 4 + 3] / 255.0;
+    dstData[dstIndex + ii * 4] = srcData[srcIndex + ii * 4] * factor;
+    dstData[dstIndex + ii * 4 + 1] = srcData[srcIndex + ii * 4 + 1] * factor;
+    dstData[dstIndex + ii * 4 + 2] = srcData[srcIndex + ii * 4 + 2] * factor;
+    dstData[dstIndex + ii * 4 + 3] = srcData[srcIndex + ii * 4 + 3];
+  }
+}
+
+function unpackPixels(srcData, width, height, imageSizes, flip_y, premultiply_alpha) {
+  var bytesPerPixel = 4;
+  var unpackedSize = width * height * bytesPerPixel;
+  var dstData = setupArrayBuffer(unpackedSize, false);
+  var srcIndex = imageSizes.skipSize;
+  var dstIndex = 0;
+  var inc = width * bytesPerPixel;
+  if (flip_y)
+    dstIndex += (height - 1) * inc;
+  for (var y = 0; y < height; ++y) {
+    copyData(srcData, srcIndex, dstData, dstIndex, width * bytesPerPixel, premultiply_alpha);
+    srcIndex += imageSizes.bytesPerRow;
+    if (flip_y)
+      dstIndex -= inc;
+    else
+      dstIndex += inc;
+  }
+  if (flip_y)
+    dstIndex += height * inc;
+  return dstData;
+}
+
+function getPixelsFromTexture2D(gl, tex, xoffset, yoffset, width, height) {
+  var fbo = gl.createFramebuffer();
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+  gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex, 0);
+  var bytesPerReadbackPixel = 4;
+  var readbackBuffer = setupArrayBuffer(width * height * bytesPerReadbackPixel, false);
+  gl.readPixels(xoffset, yoffset, width, height, gl.RGBA, gl.UNSIGNED_BYTE, readbackBuffer);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR, "read back texture pixels should succeed");
+  var bytesPerPixel = 4;
```

Comment:
Is there a strong reason for making a second copy of this data? Is the intent to bypass any pixel pack parameters, or similar? If this is just copied code from a different test then let's simplify it here.

Reply:
You are right. I copied from another test where the readbackBuffer is RGBA but the data buffer is RGB. We don't need to make another copy here. Revised.